### PR TITLE
[Wasm][AOT] Add the ability to the display the native stack frames

### PR DESCRIPTION
**Pull request by jeromelaban**
This update adds the ability to display native WebAssembly host frame, when tracing is enabled with `mono_trace_enable(TRUE);`

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/mini-exceptions.c [hunk 1]
```diff
@@ -2698,8 +2702,17 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
...
-			if (mono_ex && mono_trace_eval_exception (mono_object_class (mono_ex)))
+			if (mono_ex && mono_trace_eval_exception (mono_object_class (mono_ex))) {
 				mono_print_thread_dump_from_ctx (ctx);
+
+#ifdef HOST_WASM
+				EM_ASM(
+					var err = new Error();
+					console.log ("Native Stacktrace: \n");
+					console.log (err.stack);
+				);
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-exceptions.c:2713]: I'm not sure if we have any guidelines about Emscripten-specific stuff in the runtime like this.  I think my only hesitation would be if you were doing something that
*  needed a function from either a non-Mono-specific global JS object,
* or some local scope (ie "we know this is gonna get called from inside `library_mono.js` so we'll just refer to unqualified stuff from that scope"),
* or something that wouldn't work in a non-browser JS.

This particular usage is none of those things, so inline JS seems ok.

**Comment by jeromelaban** [comment]
Is there something specific to the PR "runtime" check that I need to be aware of, or is it just a timeout ?

**Comment by lewing** [comment]
It looks like an infrastructure issue, definitely not related to this change.

File: src/mono/mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -2698,8 +2702,17 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
 			}
 			g_print ("[%p:] EXCEPTION handling: %s.%s: %s\n", (void*)(gsize)mono_native_thread_id_get (), m_class_get_name_space (mono_object_class (obj)), m_class_get_name (mono_object_class (obj)), msg);
 			g_free (msg);
-			if (mono_ex && mono_trace_eval_exception (mono_object_class (mono_ex)))
+			if (mono_ex && mono_trace_eval_exception (mono_object_class (mono_ex))) {
 				mono_print_thread_dump_from_ctx (ctx);
+
+#ifdef HOST_WASM
+				EM_ASM(
```
Comment (lewing) [inline src/mono/mono/mini/mini-exceptions.c:2709]: Should this go into `mono_print_thread_dump_internal` instead?

File: src/mono/mono/mini/mini-exceptions.c [hunk 3]
```diff
@@ -2698,8 +2702,17 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
...
+					var err = new Error();
+					console.log ("Native Stacktrace: \n");
+					console.log (err.stack);
```
Comment (lewing) [inline src/mono/mono/mini/mini-exceptions.c:2712]: I'd rather use console.trace but a fallback might make sense

**Review by lewing — COMMENTED** [review]
We'll only reach this path at a pretty deep diagnostic level and the changes are useful

**Comment by SamMonoRT** [comment]
@lewing @radical - if the relevant changes are made, lets try to merge this.

**Comment by radical** [comment]
@jeromelaban I'm trying to see how the trace shows up in a browser. How do I reproduce this though?

**Comment by jeromelaban** [comment]
@radical it only shows up when when enabling exception tracing. I'm doing [as follows in Uno's bootstrapper](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/fa6f8617d1106e8aa63ac8e23e10980f2d66f4e7/Readme.md#aot-debugging-and-mono-tracing-net-5-only):

```
MonoInternals.mono_trace_enable(1);
MonoInternals.mono_trace_set_options("E:all");

static class MonoInternals
{
	[DllImport("__Native")]
	internal static extern void mono_trace_enable(int enable);
	[DllImport("__Native")]
	internal static extern int mono_trace_set_options(string options);
}
```

When using only dotnet, you'll need to modify the `driver.c` initialization code in order to enable it.

**Comment by SamMonoRT** [comment]
@radical @jeromelaban - is this ready to be merged ?

**Comment by jeromelaban** [comment]
@SamMonoRT on my side, it is, thanks.

**Comment by radical** [comment]
@thaystg would you be able to take a look at this?

**Comment by thaystg** [comment]
@jeromelaban thanks for your PR :), I was testing your implementation and I thought in another solution but I would like to check with you if this is enough for you. This is the PR with the change: https://github.com/dotnet/runtime/pull/49828 could you please test if this is okay for you?

This is a test with your implementation:
image

With the implementation in this new PR you will see it:
image

And if you click in the arrow you will see it:
image

**Comment by jeromelaban** [comment]
@thaystg That makes sense! Your approach avoids having to start chrome with stack dump size.

 Is it enabled using the tracing options ?

**Comment by thaystg** [comment]
Yes, the implementation is on: `mono_print_thread_dump_internal` that is called by `mono_print_thread_dump_from_ctx` that is called here: https://github.com/dotnet/runtime/blob/27fcb1ea433d0f0e2d436517d90e2a9f42e9c490/src/mono/mono/mini/mini-exceptions.c#L2706 - one line above your implementation.

**Comment by jeromelaban** [comment]
Thanks ! I'll close this PR then.

**Pull request closed without merging**